### PR TITLE
Issue 36: A troubleshooting tool built on Pravega CLI

### LINKED ISSUE #36: A troubleshooting tool built on Pravega CLI
A request to build a Pravega troubleshooting tool for stream specific metadata. The tool will look through the metadata and point out logical faults in the metadata caused due to an unknown bug or environmental fault. The implementation can be as a set of commands in the Pravega CLI.

Following the concepts in this page: https://github.com/anirudhkovuru/pravega/wiki/Pravega-Troubleshoot

### PR DESCRIPTION
**Pull request by Prabhaker24**
Purpose of the change
Fixes #36

What the code does

A troubleshoot tool using the Pravega-CLI in the form a set of 7 commands.

Each of these commands extends a base TroubleshootCommandHelper class. The TroubleshootCommandHelper class itself extends the base Command class.

The TroubleshootCommandHelper class contains the methods required to instantiate a StreamMetadataStore which can then be used to obtain access to the records required for troubleshooting.

All the commands functionality is in their execute method which is extended from the base Command class.

The various check commands are:

GeneralCheckCommand
UpdateCheckCommand
ScaleCheckCommand
CommittingTransactionsCheckCommand
TruncateCheckCommand

The above check commands implement the Check interface which contains the check method which is used by the above command's own execute methods and by the TroubleshootCheckCommand.

TroubleshootCheckCommand runs the check methods of the above check commands.

There is a Fault class to store the results. A wrapper class Record for all metadata records to aid in generating results has also been made.

utility classes have also been created. CheckUtils contains helper functions for the various troubleshoot commands and OutputUtils contains methods to convert the results into String.

Unit tests and integration tests for all the commands can be found in integrationTest.troubleshoot and unitTest.troubleshoot

How to verify it
All the unit test and integration tests should pass

### CODE REVIEW

**Comment by RaulGracia** [comment]
@Prabhaker24 what we have to do with #37? Also, the build is not passing. Please, look at this before we review this PR.

**Comment by Prabhaker24** [comment]
> @Prabhaker24 what we have to do with #37? Also, the build is not passing. Please, look at this before we review this PR.

We can ignore #37 as after the changes in the pravega codebase it's not working and it doesn't have any tests in it, sure I am looking into the build error.

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/AdminRunner.java [hunk 1]
```diff
@@ -17,11 +17,8 @@
 import io.pravega.tools.pravegacli.commands.CommandArgs;
 import io.pravega.tools.pravegacli.commands.config.ConfigListCommand;
 import io.pravega.tools.pravegacli.commands.utils.ConfigUtils;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.Scanner;
+
+import java.util.*;
```
Comment (shiveshr) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/AdminRunner.java]: nit: please revert this

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/Command.java [hunk 2]
```diff
@@ -30,13 +30,10 @@
 import io.pravega.tools.pravegacli.commands.config.ConfigListCommand;
 import io.pravega.tools.pravegacli.commands.bookkeeper.ContainerRecoverCommand;
 import io.pravega.tools.pravegacli.commands.config.ConfigSetCommand;
-import io.pravega.tools.pravegacli.commands.controller.ControllerDescribeReaderGroupCommand;
-import io.pravega.tools.pravegacli.commands.controller.ControllerDescribeScopeCommand;
-import io.pravega.tools.pravegacli.commands.controller.ControllerDescribeStreamCommand;
-import io.pravega.tools.pravegacli.commands.controller.ControllerListReaderGroupsInScopeCommand;
-import io.pravega.tools.pravegacli.commands.controller.ControllerListScopesCommand;
-import io.pravega.tools.pravegacli.commands.controller.ControllerListStreamsInScopeCommand;
+import io.pravega.tools.pravegacli.commands.controller.*;
```
Comment (shiveshr) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/Command.java:33]: nit: same here

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/troubleshoot/CommittingTransactionsCheckCommand.java [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * The method to execute the check method as part of the execution of the command.
+     */
+    @Override
+    public void execute() {
+        checkTroubleshootArgs();
+        try {
+            ScheduledExecutorService executor = getCommandArgs().getState().getExecutor();
+            store=createMetadataStore(executor);
```
Comment (shiveshr) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/troubleshoot/CommittingTransactionsCheckCommand.java]: please apply checkstyle guidelines on the code.

**Review by shiveshr — CHANGES_REQUESTED** [review]
I made one pass over the PR (only looked at code classes and not the test classes). the changes mostly look good, except few code style guidelines have not been followed.
I have pointed to a few of them. generally, please follow it on all the classes.

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/troubleshoot/CommittingTransactionsCheckCommand.java (hunk 3 above)
Comment (Prabhaker24) [reply pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/troubleshoot/CommittingTransactionsCheckCommand.java]: I have applied checkstyle guidelines

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/Command.java (hunk 2 above)
Comment (Prabhaker24) [reply pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/Command.java:33]: fixed

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/AdminRunner.java (hunk 1 above)
Comment (Prabhaker24) [reply pravega-cli/src/main/java/io/pravega/tools/pravegacli/AdminRunner.java]: fixed

**Comment by RaulGracia** [comment]
@Prabhaker24 @shiveshr is this PR still valid?

**Comment by Prabhaker24** [comment]
> @Prabhaker24 @shiveshr is this PR still valid?

Yes, @RaulGracia this is still valid I will try to get it closed by this week's end.

**Comment by RaulGracia** [comment]
@Prabhaker24 please, if this PR is still valid, let's port the changes to Pravega core repo, which is the new location of Pravega CLI.

**Pull request closed without merging**